File: src/commands/add.ts
/** dot add <path>: start tracking a file, or every file inside a directory. */

import { Task } from "../task.ts";
import {
  type DotError,
  type IoError,
  type UsageError,
  usageError,
} from "../errors.ts";
import { contractTarget, repoPathFor, resolvePath } from "../path.ts";
import {
  type Layout,
  layout,
  loadManifest,
  loadState,
  type Manifest,
  saveManifest,
  saveState,
} from "../config.ts";
import { copyFile, readBytesIfExists, sha256, stat, walkFiles } from "../fs.ts";
import { commitIfChanged, pushBestEffort } from "../git.ts";

type Added = {
  readonly repoPath: string;
  readonly target: string;
  readonly hash: string;
};

const trackOne = (
  l: Layout,
  abs: string,
): Task<Added, IoError | UsageError> => {
  const target = contractTarget(abs, l.home);
  const repoPath = repoPathFor(target);
  return copyFile(abs, l.filesDir + "/" + repoPath)
    .andThen(() => readBytesIfExists(abs))
    .andThen((bytes): Task<Added, IoError | UsageError> =>
      bytes === null
        ? Task.fail(usageError(`no such file: ${abs}`))
        : sha256(bytes).map((hash) => ({ repoPath, target, hash }))
    );
};

const listFiles = (
  l: Layout,
  abs: string,
): Task<string[], IoError | UsageError> =>
  stat(abs).andThen((info): Task<string[], IoError | UsageError> => {
    if (info === null) {
      return Task.fail(usageError(`no such path: ${abs}`));
    }
    if (abs === l.root || abs.startsWith(l.root + "/")) {
      return Task.fail(
        usageError(`cannot track dot's own data directory: ${abs}`),
      );
    }
    return info.isDirectory ? walkFiles(abs) : Task.of<string[]>([abs]);
  });

const report = (
  added: readonly Added[],
  skipped: readonly Added[],
  pushWarning: string | null,
): string =>
  [
    ...added.map((a) => `tracking ${a.target}`),
    ...skipped.map((s) => `already tracked: ${s.target}`),
    ...(added.length > 0
      ? [pushWarning ?? `committed and pushed ${added.length} file(s)`]
      : []),
  ].join("\n");

export const add = (raw: string): Task<string, DotError> =>
  Task.fromResult(layout()).andThen((l) =>
    loadManifest(l).andThen((manifest) =>
      listFiles(l, resolvePath(raw, l.home)).andThen((paths) =>
        Task.traverse(paths, (p) => trackOne(l, p)).andThen(
          (entries): Task<string, DotError> => {
            const skipped = entries.filter((e) =>
              manifest.files[e.repoPath] !== undefined
            );
            const added = entries.filter((e) =>
              manifest.files[e.repoPath] === undefined
            );
            if (added.length === 0) {
              return Task.of(report(added, skipped, null));
            }
            const next: Manifest = {
              version: 1,
              files: {
                ...manifest.files,
                ...Object.fromEntries(
                  added.map((e) => [e.repoPath, e.target] as const),
                ),
              },
            };
            return saveManifest(l, next)
              .andThen(() => loadState(l))
              .andThen((state) =>
                saveState(l, {
                  version: 1,
                  files: {
                    ...state.files,
                    ...Object.fromEntries(
                      entries.map((e) => [e.repoPath, e.hash] as const),
                    ),
                  },
                })
              )
              .andThen(() =>
                commitIfChanged(
                  l.repo,
                  `dot: add ${added.map((e) => e.target).join(", ")}`,
                )
              )
              .andThen(() => pushBestEffort(l.repo))
              .map((warning) => report(added, skipped, warning));
          },
        )
      )
    )
  );

File: src/commands/Add.scala
package dot

import cats.effect.IO
import cats.syntax.all.*
import mouse.all.*
import fs2.io.file.Path

/** dot add <path>: start tracking a file, or every file inside a directory. */

private final case class Added(repoPath: String, target: Target, hash: String)

private object Added {

  /** Copies the host file into the repo tree and records its hash. */
  def track(layout: Layout, location: Path): IO[Added] = {
    val target   = Target.contract(location, layout.home)
    val repoPath = target.repoPath
    location.copyTo(layout.repoFile(repoPath))
    *> location.sha256IfExists.flatMap:
      case None       => IO.raiseError(DotError.Usage(s"no such file: $location"))
      case Some(hash) => IO.pure(Added(repoPath, target, hash))
  }

  def report(added: List[Added], skipped: List[Added]): String = {
    val committed = added.nonEmpty.option(s"committed ${added.length} file(s) — dot sync pushes them").toList
    val tracking  = added.map(a => s"tracking ${a.target}")
    val already   = skipped.map(s => s"already tracked: ${s.target}")
    val lines     = tracking ::: already ::: committed
    lines.mkString("\n")
  }
}

extension (layout: Layout) {

  /** The files a location names — one regular file, or every file under a directory; dot's own data is refused. */
  private def filesToTrack(location: Path): IO[List[Path]] = {
    val ownData = location.startsWith(layout.root)
    location.fileKind.flatMap:
      case FileKind.Missing => IO.raiseError(DotError.Usage(s"no such path: $location"))
      case kind             =>
        val files = kind match
          case FileKind.Directory => location.walkFiles
          case _                  => IO.pure(List(location))
        IO.raiseWhen(ownData)(DotError.Usage(s"cannot track dot's own data directory: $location"))
          *> files
  }
}

/** Records the additions in manifest and state and commits them. */
private def commitAdded(layout: Layout, manifest: Manifest, entries: List[Added], added: List[Added]): IO[Unit] = {
  val files  = manifest.files ++ added.map(e => e.repoPath -> e.target)
  val hashes = entries.map(e => e.repoPath -> e.hash)
  val labels = added.map(_.target).mkString(", ")
  for
    _     <- Manifest(files).save(layout)
    state <- SyncState.load(layout)
    _     <- SyncState(state.files ++ hashes).save(layout)
    _     <- Git.in(layout.repo).commitIfChanged(s"dot: add $labels")
  yield ()
}

def add(raw: String): IO[String] = {
  for
    layout   <- Layout.resolve
    manifest <- Manifest.load(layout)
    paths    <- layout.filesToTrack(layout.locate(raw))

    entries <- paths.traverse: path =>
      Added.track(layout, path)

    (skipped, added) = entries.partition(e => manifest.files.contains(e.repoPath))

    _ <- commitAdded(layout, manifest, entries, added).unlessA(added.isEmpty)
  yield Added.report(added, skipped)
}
